dask/array/blockwise.py

```diff
@@ -1,12 +1,105 @@
 import numbers
+import string
 import warnings
 
 import tlz as toolz
 
+from collections.abc import Mapping
+from functools import reduce
+from itertools import product
+
 from .. import base, utils
 from ..delayed import unpack_collections
 from ..highlevelgraph import HighLevelGraph
-from ..blockwise import blockwise as core_blockwise
+from ..blockwise import BlockwiseIO, blockwise as core_blockwise
+
+
+class CreateArraySubgraph(Mapping):
+    """
+    Subgraph for numpy array creation utilities
+    """
+
+    def __init__(
+        self,
+        name,
+        func,
+        shape,
+        chunks,
+    ):
+        self.name = name
+        self.func = func
+        self.shape = shape
+        self.chunks = chunks
+        self._len = reduce(lambda acc, x: len(x) * acc, chunks, 1)  # Too tricky?
+
+    def __getitem__(self, key):
+        try:
+            name, *idx = key
+        except ValueError:
+            # too many / few values to unpack
+            raise KeyError(key) from None
+
+        if name != self.name:
+            raise KeyError(key)
+
+        # TODO: check for invalid indices
+
+        # Get this chunk shape
+        ndim = len(self.chunks)
+        chunk = tuple(self.chunks[i][idx[i]] for i in range(ndim))
+
+        return (self.func, chunk)
+
+    def __len__(self):
+        return self._len
+
+    def __iter__(self):
+        """
+        Produce keys for each chunk lazily.
+        """
+        ind = tuple(tuple(range(len(c))) for c in self.chunks)
+        for chunk in product(*ind):
+            yield (self.name, *chunk)
+
+
+class BlockwiseCreateArray(BlockwiseIO):
+    """
+    Specialized Blockwise Layer for numpy array creation routines.
+
+    Enables HighLevelGraph optimizations.
+    """
+
+    def __init__(
+        self,
+        name,
+        func,
+        shape,
+        chunks,
+    ):
+        self.name = name
+        self.func = func
+        self.shape = shape
+        self.chunks = chunks
+        self.io_name = "blockwise-create-" + name
+        dsk_io = CreateArraySubgraph(
+            self.io_name,
+            func,
+            shape,
+            chunks,
+        )
+        # out_ind = tuple(range(len(shape)))
+        out_ind = string.ascii_lowercase[: len(shape)]  # mayday
```

Comment:
If I could get your eyes on this @rjzamora, I'd appreciate it. Based on my reading of the docs, the `output_indices` should be able to be either a character string `"ijk"`, or a tuple `(1, 2, 3)`. Since with an ndarray we don't a priori know the number of dimensions, my intention was to do the latter. However, I ran into some issues with it that I didn't see with the string-based approach.

It seems that the difference lies here: https://github.com/dask/dask/blob/2a7ab529dc5aa5d52cca7b51d96763d5a859b55e/dask/blockwise.py#L546 
which strikes me as odd. I don't think this particular code path is exercised much, might it not be working as intended, or do I misunderstand what valid inputs should be here?

Reply:
If I use the tuple line, I get

```python
import dask.array
dask.array.ones((10,10), chunks=(1,1)).compute()
```

<details>

```pytb
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-ac15502bbc95> in <module>
      1 import dask.array
      2 
----> 3 dask.array.ones((10,10), chunks=(1,1)).compute()

~/dask/dask/dask/base.py in compute(self, **kwargs)
    278         dask.base.compute
    279         """
--> 280         (result,) = compute(self, traverse=False, **kwargs)
    281         return result
    282 

~/dask/dask/dask/base.py in compute(*args, **kwargs)
    566         postcomputes.append(x.__dask_postcompute__())
    567 
--> 568     results = schedule(dsk, keys, **kwargs)
    569     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    570 

~/dask/dask/dask/threaded.py in get(dsk, result, cache, num_workers, pool, **kwargs)
     74                 pools[thread][num_workers] = pool
     75 
---> 76     results = get_async(
     77         pool.apply_async,
     78         len(pool._pool),

~/dask/dask/dask/local.py in get_async(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)
    484                         _execute_task(task, data)  # Re-execute locally
    485                     else:
--> 486                         raise_exception(exc, tb)
    487                 res, worker_id = loads(res_info)
    488                 state["cache"][key] = res

~/dask/dask/dask/local.py in reraise(exc, tb)
    314     if exc.__traceback__ is not tb:
    315         raise exc.with_traceback(tb)
--> 316     raise exc
    317 
    318 

~/dask/dask/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    220     try:
    221         task, data = loads(task_info)
--> 222         result = _execute_task(task, data)
    223         id = get_id()
    224         result = dumps((result, id))

~/dask/dask/dask/core.py in _execute_task(arg, cache, dsk)
    119         # temporaries by their reference count and can execute certain
    120         # operations in-place.
--> 121         return func(*(_execute_task(a, cache) for a in args))
    122     elif not ishashable(arg):
    123         return arg

~/dask/dask/dask/optimization.py in __call__(self, *args)
    961         if not len(args) == len(self.inkeys):
    962             raise ValueError("Expected %d args, got %d" % (len(self.inkeys), len(args)))
--> 963         return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))
    964 
    965     def __reduce__(self):

~/dask/dask/dask/core.py in get(dsk, out, cache)
    149     for key in toposort(dsk):
    150         task = dsk[key]
--> 151         result = _execute_task(task, cache)
    152         cache[key] = result
    153     result = _execute_task(out, cache)

~/dask/dask/dask/core.py in _execute_task(arg, cache, dsk)
    119         # temporaries by their reference count and can execute certain
    120         # operations in-place.
--> 121         return func(*(_execute_task(a, cache) for a in args))
    122     elif not ishashable(arg):
    123         return arg

TypeError: __call__() takes 2 positional arguments but 3 were given
```

</details>
suggesting they are unexpectedly unpacked.

Reply:
>which strikes me as odd. I don't think this particular code path is exercised much, might it not be working as intended, or do I misunderstand what valid inputs should be here?

You are right to question this :)

This is a new piece of logic that has only been tested for the few DataFrame-focused cases where `BlockwiseIO` is used.  It certainly seems that we need to consider the case that `output_indices` is a tuple - Sorry about that!

Reply:
Thanks for pointing this out - If I understand correctly, there is no reason to use the dimensions here to define `dsk`.  For, `BlockwiseIO` there should always be a single "dummy" variable (which will be replaced with the packed args for the IO function at run time).  I submitted #6933 with the simple fix.

Reply:
Thanks for looking into it @rjzamora!